platforms/software/dependency-management/src/main/java/org/gradle/internal/resolve/resolver/ExcludingVariantArtifactSet.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.resolve.resolver;
+
+import org.gradle.api.artifacts.ModuleIdentifier;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ResolvableArtifact;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ResolvedArtifactSet;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ResolvedVariant;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.excludes.specs.ExcludeSpec;
+import org.gradle.api.internal.attributes.ImmutableAttributes;
+import org.gradle.internal.Describables;
+import org.gradle.internal.DisplayName;
+import org.gradle.internal.component.external.model.ImmutableCapabilities;
+import org.gradle.internal.component.model.VariantResolveMetadata;
+
+import javax.annotation.Nullable;
+import java.util.Objects;
+
+/**
+ * A {@link ResolvedVariant} that applies artifact exclusions to a delegate {@link ResolvedVariant}.
+ */
+public class ExcludingVariantArtifactSet implements ResolvedVariant, VariantResolveMetadata.Identifier {
```

Comment:
🤔 Even though there are multiple implementations of `ResolvedVariant`, it seems like the only one we're interested in wrapping in this decorator is `ArtifactBackedResolvedVariant`.  So why not have `ExcludingVariantArtifactSet` just subtype that and avoid allocating another instance?

This would simplify the logic in `VariantResolvingArtifactSet` and restore the use of a single loop.  It would trim the size of this type significantly.

In general, I think we're too quick to delegate instead of subtype throughout Gradle, and it leads to more complex and less comprehensible code.

Reply:
Lets leave this as-is